lib/src/spec/message.dart

```diff
@@ -49,38 +50,71 @@ class MessageData<T> {
   }
 }
 
+/// Delta extension configuration for [MessageExtras]
+class DeltaExtras {
+  /// the id of the message the delta was generated from
+  final String from;
+
+  /// the delta format. Only "vcdiff" is supported currently
+  final String format;
+
+  /// initializes with from and format
+  const DeltaExtras({this.from, this.format});
+
+  /// create instance from a map
+  static DeltaExtras _fromMap(Map value) {
+    if (value == null) return null;
+    return DeltaExtras(
+      from: value['from'] as String,
+      format: value['format'] as String,
+    );
+  }
+
+  @override
+  bool operator ==(Object other) =>
+      other is DeltaExtras && other.from == from && other.format == format;
+
+  @override
+  int get hashCode => '$from:$format'.hashCode;
+}
+
 /// Handles supported message extras types, their encoding and decoding
-class MessageExtras<T> {
-  final T _extras;
+class MessageExtras {
+  /// json-encodable map of extras
+  final Map<String, dynamic> extras;
+
+  /// configuration for delta compression extension
+  DeltaExtras _delta;
 
-  /// Only Map and List types are supported
-  MessageExtras(this._extras) : assert(T == Map);
+  /// delta configuration received from channel message
+  DeltaExtras get delta => _delta;
 
-  /// retrieve extras
-  T get extras => _extras;
+  /// extras is a mandatory json-encodable argument while delta is an extension
+  ///  which defines configuration for delta compression extension
+  MessageExtras(this.extras);
 
   /// initializes [MessageExtras] with given value and validates
   /// the data type, runtime
-  static MessageExtras fromValue(Object value) {
-    if (value == null) {
-      return null;
-    }
-    assert(
-      value is MessageExtras || value is Map,
-      'Message extras must be either `Map`, or `null`.'
-      ' Does not support $value ("${value.runtimeType}")',
-    );
-    if (value is MessageExtras) {
-      return value;
-    } else if (value is Map) {
-      return MessageExtras<Map>(value);
-    } else {
-      throw AssertionError(
-        'Message extras must be either `Map`, or `null`.'
-        ' Does not support $value ("${value.runtimeType}")',
-      );
-    }
+  static MessageExtras fromMap(Map<String, dynamic> value) {
+    if (value == null) return null;
+    value = Map<String, dynamic>.from(value);
+    final deltaMap = value.remove('delta') as Map;
+    final extras = MessageExtras(value)
+      .._delta = DeltaExtras._fromMap(deltaMap);
```

Comment:
I'm not convinced this is appropriate use of cascades, as it makes this code very confusing to the casual reader, especially given you're mutating a member with an underscore prefix.

Reply:
@QuintinWillison what do you suggest is better here? Our lint rules enforce cascading wherever applicable.

One option is to return it without assigning to a variable `extras` which will then look like:
```dart
return MessageExtras(value).._delta = DeltaExtras._fromMap(deltaMap);
```

Reply:
Why not make it optional argument on the `MessageExtras` initialiser?

Reply:
Considering inputs from both: d30ed0ea880c155e6f5ff7290c4974c592b80e1b

> You may wonder why a private named constructor instead of adding to default constructor:
_Deltas is not supposed to be assigned by SDK users in any way as it is read only by definition.
However, it still stands as an exposed API via `MessageExtras.fromMap` due to technical limitations of ability to configure a package private constructor. So, just trying to reduce the exposure as much as we can._